File: dc_general_ability/src/main/scala/com/hopson/dc/ability/scala/hbase/HbaseFactory.scala
package com.hopson.dc.ability.scala.hbase

import com.hopson.dc.common.config.Config
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.hbase.{HBaseConfiguration, TableName}
import org.apache.hadoop.hbase.client.{Admin, Connection, ConnectionFactory, Table}

/**
 * @Author: xiaozhennan
 * @Date: 2020/1/13 11:51
 * @Package: com.hopson.dc.ability.scala.hbase
 * @ClassName: HbaseFactory
 * @Description:
 **/
object HbaseFactory {

  private val instance: HbaseFactory = new HbaseFactory;
  private var init_flag = false;

  def getInstance(): HbaseFactory = {
    if (init_flag == false) {
      println("hbase没有初始化!")
      throw new RuntimeException
    }
    instance
  }

  def build(config: Config) = {
    if (!init_flag) {
      val zookeeperStr = config.get("hbase.zookeeper.server")
      val zookeeperPort = config.get("hbase.zookeeper.port")
      instance.hbasecfg.set("hbase.zookeeper.quorum", zookeeperStr)
      instance.hbasecfg.set("hbase.zookeeper.property.clientPort", zookeeperPort)
      instance.conn = ConnectionFactory.createConnection(instance.hbasecfg)
      instance.admin = instance.conn.getAdmin
      init_flag = true
    }
  }


}

class HbaseFactory private {


  private var conn: Connection = null
  private var admin: Admin = null
  private var hbasecfg: Configuration = HBaseConfiguration.create

  def getConnection(): Connection = {
    return this.conn;
  }

  def getAdmin(): Admin = {
    return this.admin;
  }

  def getTable(tableName: String): Table = {
    val table = conn.getTable(TableName.valueOf(tableName));
    return table;
  }

}

File: dc_general_ability/src/main/java/com/hopson/dc/ability/java/hbase/HbaseFactory.java
package com.hopson.dc.ability.java.hbase;

import com.hopson.dc.common.config.Config;
import com.hopson.dc.common.config.ConfigHolder;
import com.hopson.dc.common.config.ConfigStore;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.hbase.HBaseConfiguration;
import org.apache.hadoop.hbase.client.Admin;
import org.apache.hadoop.hbase.client.Connection;
import org.apache.hadoop.hbase.client.ConnectionFactory;

import java.io.IOException;


/**
 * @Date: 2018/8/23 11:00
 * @Author: xiaozhennan
 * @Package: com.ipaynow.dc.srv.factory
 * @ClassName: HbaseFactory
 * @Description: 访问Hbase单例工程, 保存Hbase链接
 **/
public class HbaseFactory {


    private static volatile HbaseFactory instance = null;

    private HbaseFactory() {
    }

    private Configuration hbaseCfg = null;
    private Config config = null;
    private Connection conn;

    private Admin hbaseAdmin;
    private boolean init_flag = false;

    public static HbaseFactory getInstance() {
        if (instance != null) {
            return instance;
        }
        return null;
    }

    public static void build(Config cfg) {
        if (null == instance) {
            synchronized (HbaseFactory.class) {
                instance = new HbaseFactory();
                instance.config = cfg;
                try {
                    instance.init();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    private void init() throws IOException {
        hbaseCfg = HBaseConfiguration.create();
        ConfigStore cache = ConfigHolder.getCache();
        String zookeeperStr = config.get("hbase.zookeeper.server");
        String zookeeperPort = config.get("hbase.zookeeper.port");
        hbaseCfg.set("hbase.zookeeper.quorum", zookeeperStr);
        hbaseCfg.set("hbase.zookeeper.property.clientPort", zookeeperPort);
        conn = ConnectionFactory.createConnection(hbaseCfg);
        hbaseAdmin = conn.getAdmin();
        this.init_flag = true;
    }

    public Connection getConnection() {
        return conn;
    }

    public Admin getHbaseAdmin() {
        return hbaseAdmin;
    }

}
